# 01 · Data Cleaning and Preparation

**Project:** DataCo Supply Chain Analytics
**Dataset:** [DataCo Smart Supply Chain for Big Data Analysis](https://www.kaggle.com/datasets/shashwatwork/dataco-smart-supply-chain-for-big-data-analysis) (Kaggle), 180,519 order items, January 2015 – January 2018
**Input:** `data/raw/DataCoSupplyChainDataset.csv`

## Purpose
Turn the raw file into clean, documented data that every later analysis relies on.

## Outputs
1. `data/clean/order_items_clean.csv`: one row per order item
2. `data/clean/orders_clean.csv`: one row per order, used for delivery and fraud analysis to avoid double-counting multi-item orders
3. `reports/data_quality_log.md`: every issue found and the action taken

## Contents
- Phase 2 check: validate the metric definitions against the raw data
- Phase 3A: profile and verify
- Phase 3B: drop, rename, fix types and clean text
- Phase 3C: derived columns, order-level table and saved outputs

Metric definitions are in [docs/metric_definitions.md](../docs/metric_definitions.md).

In [1]:
import pandas as pd
df = pd.read_csv('../data/raw/Rawdata_DataCo_Supply_Chain.csv', encoding='latin-1')
df.shape

(180519, 53)

## Phase 2: metric definitions

Before cleaning the data, I confirmed that each metric is defined correctly
and consistently with the project brief. The code below checks:

- that each metric's formula matches its documented definition
- that the units and time windows are consistent
- how edge cases (nulls, zeros, duplicates) are handled

This keeps any cleaning steps from changing what a metric means.


In [2]:
import pandas as pd

df = pd.read_csv('../data/raw/Rawdata_DataCo_Supply_Chain.csv', encoding='latin-1')

# Volume
print('Order items:', len(df))
print('Orders:', df['Order Id'].nunique())
print('Customers:', df['Customer Id'].nunique())
print('Products:', df['Product Card Id'].nunique())

# Financial
revenue = df['Order Item Total'].sum()
profit = df['Order Profit Per Order'].sum()
print(f'Revenue: ${revenue:,.0f}  Profit: ${profit:,.0f}  Margin: {profit/revenue:.1%}')
print(f"Loss item rate: {(df['Order Profit Per Order'] < 0).mean():.1%}")

# Delivery (order level, shipped only)
orders = df.drop_duplicates('Order Id')
shipped = orders[orders['Delivery Status'] != 'Shipping canceled']
print(f"Shipped orders: {len(shipped):,}  Late rate: {shipped['Late_delivery_risk'].mean():.1%}")

# Fraud (order level)
print(f"Fraud orders: {(orders['Order Status'] == 'SUSPECTED_FRAUD').sum():,}")

Order items: 180519
Orders: 65752
Customers: 20652
Products: 118
Revenue: $33,054,402  Profit: $3,966,903  Margin: 12.0%
Loss item rate: 18.7%
Shipped orders: 62,897  Late rate: 57.3%
Fraud orders: 1,488


# Phase 3: Data cleaning
Goal: produce two clean tables (order items and orders) and a data quality log recording every decision.

## 3.1 Setup

In [3]:
import pandas as pd
import numpy as np
from pathlib import Path

pd.set_option('display.width', 200, 'display.max_columns', 60, 'display.max_colwidth', 45)
RAW = Path('../data/raw/Rawdata_DataCo_Supply_Chain.csv')
OUT = Path('../data/clean')
OUT.mkdir(parents=True, exist_ok=True)

# Data quality log: one entry per issue found
log = []
def record(issue, columns, rows, action):
    log.append({'issue': issue, 'columns': columns, 'rows_affected': rows, 'action': action})

## 3.2 Load the data
The file is not UTF-8, so it needs `encoding='latin-1'`.

In [4]:
df = pd.read_csv(RAW, encoding='latin-1')
print(f'{df.shape[0]:,} rows x {df.shape[1]} columns')
df.head(3)

180,519 rows x 53 columns


,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Cally,20755,Holloway,XXXXXXXXX,Consumer,PR,5365 Noble Nectar Island,725.0,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,1/31/2018 22:56,77202,1360,13.110000,0.04,180517,327.75,0.29,1,327.75,314.640015,91.250000,Southeast Asia,Java Occidental,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Irene,19492,Luna,XXXXXXXXX,Consumer,PR,2679 Rustic Loop,725.0,2,Fitness,18.279451,-66.037064,Pacific Asia,Bikaner,India,19492,1/13/2018 12:27,75939,1360,16.389999,0.05,179254,327.75,-0.80,1,327.75,311.359985,-249.089996,South Asia,Rajastán,PENDING,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,EE. UU.,XXXXXXXXX,Gillian,19491,Maldonado,XXXXXXXXX,Consumer,CA,8510 Round Bear Gate,95125.0,2,Fitness,37.292233,-121.881279,Pacific Asia,Bikaner,India,19491,1/13/2018 12:06,75938,1360,18.030001,0.06,179253,327.75,-0.80,1,327.75,309.720001,-247.779999,South Asia,Rajastán,CLOSED,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class


## 3.3 Profile every column
Data type, null count, share of nulls, distinct values and a sample value for each column.

In [5]:
profile = pd.DataFrame({
    'dtype': df.dtypes.astype(str),
    'nulls': df.isna().sum(),
    'null_%': (df.isna().mean() * 100).round(1),
    'distinct': df.nunique(),
    'sample': df.iloc[0],
})
profile

,dtype,nulls,null_%,distinct,sample
Type,object,0,0.0,4,DEBIT
Days for shipping (real),int64,0,0.0,7,3
Days for shipment (scheduled),int64,0,0.0,4,4
Benefit per order,float64,0,0.0,21998,91.25
Sales per customer,float64,0,0.0,2927,314.640015
Delivery Status,object,0,0.0,4,Advance shipping
Late_delivery_risk,int64,0,0.0,2,0
Category Id,int64,0,0.0,51,73
Category Name,object,0,0.0,50,Sporting Goods
Customer City,object,0,0.0,563,Caguas


**Findings from the profile**
- Product Description is completely empty and Order Zipcode is about 86% empty.
- Customer Email and Customer Password contain one masked value, and Product Status has a single value, so they carry no information.
- Several column pairs have identical distinct counts (e.g. Customer Id and Order Customer Id), which suggests duplicates. These are tested in 3.4 rather than assumed.

## 3.4 Verify suspected duplicate columns
A column is dropped only if its values are identical to another column on every row.

In [6]:
pairs = [('Benefit per order', 'Order Profit Per Order'),
         ('Sales per customer', 'Order Item Total'),
         ('Order Customer Id', 'Customer Id'),
         ('Order Item Cardprod Id', 'Product Card Id'),
         ('Product Category Id', 'Category Id'),
         ('Product Price', 'Order Item Product Price')]

pairs = [(a, b) for a, b in pairs if a in df.columns and b in df.columns]
pd.DataFrame([(a, b, (df[a] == df[b]).all()) for a, b in pairs],
             columns=['column', 'duplicate_of', 'identical'])

,column,duplicate_of,identical
0,Benefit per order,Order Profit Per Order,True
1,Sales per customer,Order Item Total,True
2,Order Customer Id,Customer Id,True
3,Order Item Cardprod Id,Product Card Id,True
4,Product Category Id,Category Id,True
5,Product Price,Order Item Product Price,True


## 3.5 Verify how the financial columns relate
This decides which column is used as revenue and what the profit column really measures.

In [7]:
checks = {
    'Sales = Product Price x Quantity':
        np.isclose(df['Order Item Product Price'] * df['Order Item Quantity'], df['Sales'], atol=0.02).mean(),
    'Order Item Total = Sales - Discount':
        np.isclose(df['Sales'] - df['Order Item Discount'], df['Order Item Total'], atol=0.02).mean(),
    'Profit Ratio = Profit / Order Item Total':
        np.isclose(df['Order Profit Per Order'] / df['Order Item Total'], df['Order Item Profit Ratio'], atol=0.011).mean(),
    'Order Item Id is unique': float(df['Order Item Id'].is_unique),
    'Fully duplicated rows': df.duplicated().sum(),
}
pd.Series(checks, name='share of rows passing').to_frame()

,share of rows passing
Sales = Product Price x Quantity,1.0
Order Item Total = Sales - Discount,1.0
Profit Ratio = Profit / Order Item Total,1.0
Order Item Id is unique,1.0
Fully duplicated rows,0.0


**Decisions**
- Sales is the amount before discount; Order Item Total is after discount, so revenue = Order Item Total.
- Order Profit Per Order is item-level profit despite its name, because it matches each item's profit ratio.
- Order Item Id is unique, so each row is one order item.

## 3.6 Check order-level consistency and status fields
If order details never change within an order, an order-level table can safely be built later.

In [8]:
df.groupby('Order Id')[['Order Status', 'Type', 'Shipping Mode', 'Delivery Status',
                        'Market', 'Order Region', 'Customer Id']].nunique().max()

Order Status       1
Type               1
Shipping Mode      1
Delivery Status    1
Market             1
Order Region       1
Customer Id        1
dtype: int64

In [9]:
pd.crosstab(df['Order Status'], df['Delivery Status'])

Delivery Status,Advance shipping,Late delivery,Shipping canceled,Shipping on time
Order Status,,,,
CANCELED,0,0,3692,0
CLOSED,4809,11109,0,3698
COMPLETE,14136,34199,0,11156
ON_HOLD,2413,5450,0,1941
PAYMENT_REVIEW,420,1082,0,391
PENDING,4857,11712,0,3658
PENDING_PAYMENT,9588,22922,0,7322
PROCESSING,5369,12503,0,4030
SUSPECTED_FRAUD,0,0,4062,0


**Findings**
- Every suspected fraud and cancelled order has Delivery Status "Shipping canceled". These orders never shipped, so they are excluded from the late-delivery analysis. Delivery Status is also excluded from the fraud model, because it is only known after the fraud decision (data leakage).
- Payment type fully determines order status, and all fraud is paid by bank transfer. Real data is rarely this tidy, which suggests the dataset is simulated.

In [10]:
record('Fraud and cancelled orders never shipped', 'Order Status, Delivery Status',
       int(df['Delivery Status'].eq('Shipping canceled').sum()),
       'Excluded from delivery analysis; Delivery Status excluded from fraud model (leakage)')
record('Payment type fully determines order status', 'Type, Order Status', len(df),
       'Kept; noted as synthetic-data limitation')
print(len(log), 'log entries so far')

2 log entries so far


## 3.7 Drop columns that add nothing
Each column is dropped for a documented reason: masked, empty, personal, too detailed, or a proven duplicate (3.4). Category Id and Department Id are kept because the star schema uses them as keys.

In [11]:
drop_cols = {
    'Customer Email': 'masked (single value)',
    'Customer Password': 'masked (single value)',
    'Customer Fname': 'personal data not needed (UK GDPR data minimisation)',
    'Customer Lname': 'personal data not needed (UK GDPR data minimisation)',
    'Customer Street': 'personal data not needed (UK GDPR data minimisation)',
    'Customer Zipcode': 'too detailed; market and region are used',
    'Product Description': '100% empty',
    'Product Image': 'image URL only',
    'Product Status': 'single value (0)',
    'Order Zipcode': 'about 86% empty',
    'Latitude': 'too detailed; market and region are used',
    'Longitude': 'too detailed; market and region are used',
    'Benefit per order': 'duplicate of Order Profit Per Order',
    'Sales per customer': 'duplicate of Order Item Total',
    'Order Customer Id': 'duplicate of Customer Id',
    'Order Item Cardprod Id': 'duplicate of Product Card Id',
    'Product Category Id': 'duplicate of Category Id',
    'Product Price': 'duplicate of Order Item Product Price',
}

present = [c for c in drop_cols if c in df.columns]
for col in present:
    record('Column not needed', col, len(df), f'Dropped: {drop_cols[col]}')

df = df.drop(columns=present)
print(f'Dropped {len(present)} columns; {df.shape[1]} remain')

Dropped 18 columns; 35 remain


## 3.8 Rename columns to snake_case
Short, lowercase names without spaces are easier to use in Python, SQL and Power BI. Two names are corrected to say what the column really holds: Order Profit Per Order becomes item_profit (it is item-level, see 3.5), and Order Item Total becomes revenue.

In [12]:
rename = {
    'Type': 'payment_type',
    'Days for shipping (real)': 'days_shipping_real',
    'Days for shipment (scheduled)': 'days_shipping_scheduled',
    'Delivery Status': 'delivery_status',
    'Late_delivery_risk': 'late_delivery_flag',
    'Category Id': 'category_id',
    'Category Name': 'category_name',
    'Customer City': 'customer_city',
    'Customer Country': 'customer_country',
    'Customer Id': 'customer_id',
    'Customer Segment': 'customer_segment',
    'Customer State': 'customer_state',
    'Department Id': 'department_id',
    'Department Name': 'department_name',
    'Market': 'market',
    'Order City': 'order_city',
    'Order Country': 'order_country',
    'order date (DateOrders)': 'order_date',
    'Order Id': 'order_id',
    'Order Item Discount': 'discount_amount',
    'Order Item Discount Rate': 'discount_rate',
    'Order Item Id': 'order_item_id',
    'Order Item Product Price': 'unit_price',
    'Order Item Profit Ratio': 'profit_ratio',
    'Order Item Quantity': 'quantity',
    'Sales': 'gross_sales',
    'Order Item Total': 'revenue',
    'Order Profit Per Order': 'item_profit',
    'Order Region': 'order_region',
    'Order State': 'order_state',
    'Order Status': 'order_status',
    'Product Card Id': 'product_id',
    'Product Name': 'product_name',
    'shipping date (DateOrders)': 'ship_date',
    'Shipping Mode': 'shipping_mode',
}
df = df.rename(columns=rename)
print(df.columns.tolist())

['payment_type', 'days_shipping_real', 'days_shipping_scheduled', 'delivery_status', 'late_delivery_flag', 'category_id', 'category_name', 'customer_city', 'customer_country', 'customer_id', 'customer_segment', 'customer_state', 'department_id', 'department_name', 'market', 'order_city', 'order_country', 'order_date', 'order_id', 'discount_amount', 'discount_rate', 'order_item_id', 'unit_price', 'profit_ratio', 'quantity', 'gross_sales', 'revenue', 'item_profit', 'order_region', 'order_state', 'order_status', 'product_id', 'product_name', 'ship_date', 'shipping_mode']


## 3.9 Fix data types
Both date columns are stored as text. They are converted to datetime so dates can be compared, grouped by month and used to calculate time differences.

In [13]:
print('Before:', df['order_date'].dtype, '| example:', df['order_date'].iloc[0])

df['order_date'] = pd.to_datetime(df['order_date'], format='%m/%d/%Y %H:%M')
df['ship_date'] = pd.to_datetime(df['ship_date'], format='%m/%d/%Y %H:%M')
record('Dates stored as text', 'order_date, ship_date', len(df), 'Converted to datetime (format m/d/Y H:M)')

print('After:', df['order_date'].dtype)
print('Order dates run from', df['order_date'].min().date(), 'to', df['order_date'].max().date())

Before: object | example: 1/31/2018 22:56
After: datetime64[ns]
Order dates run from 2015-01-01 to 2018-01-31


In [14]:
gap = (df['ship_date'] - df['order_date']).dt.days - df['days_shipping_real']

print('Ship date before order date:', (df['ship_date'] < df['order_date']).sum())
print('Rows where the date difference disagrees with days_shipping_real:', (gap != 0).sum())
print('Size of the disagreement (days):', gap[gap != 0].unique())

record('Date difference disagrees with days_shipping_real by 1 day', 'ship_date, days_shipping_real',
       int((gap != 0).sum()),
       'Kept days_shipping_real as the delivery measure; the gap is a time-of-day rounding effect')

Ship date before order date: 0
Rows where the date difference disagrees with days_shipping_real: 4657
Size of the disagreement (days): [-1]


**Why the 1-day gap is not a problem:** `.dt.days` only counts complete 24-hour periods, so an order placed at 22:00 and shipped two calendar days later at 09:00 counts as 1 day, not 2. days_shipping_real counts calendar days, so it is used as the delivery measure throughout.

## 3.10 Standardise text
Some labels have trailing or double spaces (e.g. "South of  USA "), which would split one region into two groups. Customer country uses a Spanish abbreviation for the United States.

In [15]:
text_cols = df.select_dtypes(include=['object', 'string']).columns
before = {c: set(df[c].dropna().unique()) for c in text_cols}

for c in text_cols:
    df[c] = df[c].str.strip().str.replace(r'\s{2,}', ' ', regex=True)

changed = {c: sorted(before[c] - set(df[c].dropna().unique()))[:4] for c in text_cols
           if before[c] != set(df[c].dropna().unique())}
for c, examples in changed.items():
    print(f'{c}: fixed {examples}')

record('Trailing or double spaces in labels', ', '.join(changed), len(df),
       'Stripped and collapsed whitespace (e.g. "South of  USA " to "South of USA")')

category_name: fixed ['As Seen on  TV!', 'Baby ', 'Books ', 'CDs ']
department_name: fixed ['Health and Beauty ']
order_region: fixed ['South of  USA ', 'US Center ', 'West of USA ']
product_name: fixed ['DVDs ', 'Smart watch ', 'Sports Books ', 'Toys ']


In [16]:
print('Before:', df['customer_country'].value_counts().to_dict())
df['customer_country'] = df['customer_country'].replace({'EE. UU.': 'United States'})
print('After: ', df['customer_country'].value_counts().to_dict())

record('Customer country in Spanish', 'customer_country',
       int((df['customer_country'] == 'United States').sum()), 'Mapped "EE. UU." to "United States"')
record('Order country names in Spanish (e.g. "Francia")', 'order_country', len(df),
       'Kept as-is; analysis uses English market and order_region. Translate before building a map visual')

Before: {'EE. UU.': 111146, 'Puerto Rico': 69373}
After:  {'United States': 111146, 'Puerto Rico': 69373}


**Order country is left in Spanish** (e.g. "Estados Unidos", "Francia"). Translating 164 countries adds little, because the analysis groups by the English market and order_region columns. It would only matter for a map visual.

## 3.11 Missing values and outliers

In [17]:
print('Missing values remaining:', int(df.isna().sum().sum()))
df[['quantity', 'unit_price', 'discount_rate', 'revenue', 'item_profit', 'profit_ratio',
    'days_shipping_real', 'days_shipping_scheduled']].describe().round(2).T

Missing values remaining: 0


,count,mean,std,min,25%,50%,75%,max
quantity,180519.0,2.13,1.45,1.00,1.00,1.00,3.00,5.00
unit_price,180519.0,141.23,139.73,9.99,50.00,59.99,199.99,1999.99
discount_rate,180519.0,0.10,0.07,0.00,0.04,0.10,0.16,0.25
revenue,180519.0,183.11,120.04,7.49,104.38,163.99,247.40,1939.99
item_profit,180519.0,21.97,104.43,-4274.98,7.00,31.52,64.80,911.80
profit_ratio,180519.0,0.12,0.47,-2.75,0.08,0.27,0.36,0.50
days_shipping_real,180519.0,3.50,1.62,0.00,2.00,3.00,5.00,6.00
days_shipping_scheduled,180519.0,2.93,1.37,0.00,2.00,4.00,4.00,4.00


In [18]:
loss_rows = int((df['item_profit'] < 0).sum())
print(f'Loss-making items: {loss_rows:,} ({loss_rows / len(df):.1%})')
print('Worst profit ratio:', df['profit_ratio'].min())

record('Negative item profit (down to -275% of revenue)', 'item_profit, profit_ratio', loss_rows,
       'Kept: genuine losses, central to Q2; flagged later with is_loss')

Loss-making items: 33,784 (18.7%)
Worst profit ratio: -2.75


In [19]:
df['discount_rate'].value_counts().sort_index()

discount_rate
0.00    10028
0.01    10028
0.02    10028
0.03    10029
0.04    10029
0.05    10029
0.06    10029
0.07    10029
0.09    10029
0.10    10029
0.12    10029
0.13    10029
0.15    10029
0.16    10029
0.17    10029
0.18    10029
0.20    10029
0.25    10029
Name: count, dtype: int64

In [20]:
record('Discount rates evenly spread (about 10,029 items each)', 'discount_rate', len(df),
       'Kept; noted as synthetic-data limitation affecting discount and volume conclusions')

In [21]:
monthly = df.groupby(df['order_date'].dt.to_period('M')).agg(
    orders=('order_id', 'nunique'), products=('product_name', 'nunique'))
monthly.tail(8)

,orders,products
order_date,,
2017-06,1676,42
2017-07,1776,42
2017-08,1768,42
2017-09,1723,53
2017-10,2101,20
2017-11,2055,8
2017-12,2124,14
2018-01,2123,10


In [22]:
record('Product range shrinks from Q4 2017 and data ends Jan 2018', 'product_name, order_date',
       int((df['order_date'] >= '2017-10-01').sum()),
       'Kept; calendar 2017 used as the last full year; trends after Sep 2017 treated with caution')

print(f'{df.shape[0]:,} rows x {df.shape[1]} columns | {len(log)} log entries')

180,519 rows x 35 columns | 28 log entries


**Summary of Part B**
- 18 columns dropped with documented reasons; 35 remain, renamed to snake_case.
- Dates converted; days_shipping_real kept as the delivery measure.
- Label whitespace cleaned; customer country standardised.
- No missing values remain. Negative profits are real losses and kept.
- Three signs of simulated data are now logged: payment type determines status, discount rates are evenly spread, and the product range shrinks from Q4 2017.

##3.12 Add derived columns
Helper columns are calculated once during cleaning so every later analysis uses the same definitions

In [23]:
# Date parts: .dt gives access to parts of a date column
df['order_year']    = df['order_date'].dt.year                               # e.g. 2017
df['order_month']   = df['order_date'].dt.to_period('M').dt.to_timestamp()  # first day of the month, e.g. 2017-03-01
df['order_weekday'] = df['order_date'].dt.day_name()                         # e.g. 'Monday'

# Number of days late (positive) or early (negative)
df['shipping_delay'] = df['days_shipping_real'] - df['days_shipping_scheduled']

# True/False flags: a comparison gives True or False on every row
df['is_shipped']   = df['delivery_status'] != 'Shipping canceled'   # != means "is not"
df['is_late']      = df['late_delivery_flag'].astype(bool)          # turns 1/0 into True/False
df['is_loss']      = df['item_profit'] < 0
df['is_fraud']     = df['order_status'] == 'SUSPECTED_FRAUD'        # == means "is equal to"
df['is_cancelled'] = df['order_status'] == 'CANCELED'

# Group discount rates into 4 bands
df['discount_band'] = pd.cut(df['discount_rate'],
                             bins=[-0.001, 0.05, 0.10, 0.15, 1],
                             labels=['0-5%', '6-10%', '11-15%', '16-25%'])

print(df.shape[1], 'columns')

45 columns


In [24]:
shipped_items = df[df['is_shipped']]      # keep only rows where is_shipped is True
print('is_late matches shipping_delay > 0:', ((shipped_items['shipping_delay'] > 0) == shipped_items['is_late']).mean())

print(df[['is_shipped', 'is_late', 'is_loss', 'is_fraud', 'is_cancelled']].sum())   # .sum() counts the Trues
print(df['discount_band'].value_counts().sort_index())

is_late matches shipping_delay > 0: 1.0
is_shipped      172765
is_late          98977
is_loss          33784
is_fraud          4062
is_cancelled      3692
dtype: int64
discount_band
0-5%      50142
6-10%     40116
11-15%    30087
16-25%    60174
Name: count, dtype: int64


In [25]:
df['order_weekday'].value_counts()

order_weekday
Sunday       25946
Friday       25925
Saturday     25901
Monday       25786
Thursday     25752
Tuesday      25622
Wednesday    25587
Name: count, dtype: int64

In [ ]:
## 3.13 Build the order-level table
One row per order, used for delivery and fraud analysis. Order details are the same for every item in an order (checked in 3.6), so the first value is kept. Money and quantities are added up.

In [26]:
orders = df.groupby('order_id').agg(
    # Same for every item in the order: keep the first value
    order_date=('order_date', 'first'),
    order_year=('order_year', 'first'),
    order_month=('order_month', 'first'),
    customer_id=('customer_id', 'first'),
    customer_segment=('customer_segment', 'first'),
    customer_country=('customer_country', 'first'),
    payment_type=('payment_type', 'first'),
    order_status=('order_status', 'first'),
    delivery_status=('delivery_status', 'first'),
    shipping_mode=('shipping_mode', 'first'),
    market=('market', 'first'),
    order_region=('order_region', 'first'),
    order_country=('order_country', 'first'),
    days_shipping_real=('days_shipping_real', 'first'),
    days_shipping_scheduled=('days_shipping_scheduled', 'first'),
    shipping_delay=('shipping_delay', 'first'),
    is_shipped=('is_shipped', 'first'),
    is_late=('is_late', 'first'),
    is_fraud=('is_fraud', 'first'),
    is_cancelled=('is_cancelled', 'first'),
    # Different for each item: count or add up
    n_items=('order_item_id', 'size'),          # how many items
    total_quantity=('quantity', 'sum'),
    gross_sales=('gross_sales', 'sum'),
    discount_amount=('discount_amount', 'sum'),
    revenue=('revenue', 'sum'),
    profit=('item_profit', 'sum'),
    n_categories=('category_name', 'nunique'),  # how many different categories
).reset_index()

orders['avg_discount_rate'] = orders['discount_amount'] / orders['gross_sales']

print(f'{len(orders):,} orders x {orders.shape[1]} columns')
orders.head(3)

65,752 orders x 29 columns


,order_id,order_date,order_year,order_month,customer_id,customer_segment,customer_country,payment_type,order_status,delivery_status,shipping_mode,market,order_region,order_country,days_shipping_real,days_shipping_scheduled,shipping_delay,is_shipped,is_late,is_fraud,is_cancelled,n_items,total_quantity,gross_sales,discount_amount,revenue,profit,n_categories,avg_discount_rate
0,1,2015-01-01 00:00:00,2015,2015-01-01,11599,Consumer,United States,CASH,CLOSED,Advance shipping,Standard Class,LATAM,Central America,México,2,4,-2,True,False,False,False,1,1,299.980011,60.00,239.979996,88.790001,1,0.200013
1,2,2015-01-01 00:21:00,2015,2015-01-01,256,Consumer,United States,PAYMENT,PENDING_PAYMENT,Advance shipping,Standard Class,LATAM,South America,Colombia,3,4,-1,True,False,False,False,3,7,579.980011,50.60,529.380005,195.900002,3,0.087244
2,4,2015-01-01 01:03:00,2015,2015-01-01,8827,Home Office,United States,CASH,CLOSED,Late delivery,Standard Class,LATAM,South America,Colombia,5,4,1,True,True,False,False,4,14,699.850010,78.98,620.870014,124.090000,4,0.112853


### Checking cells for reconciliation

In [27]:
# assert does nothing if the statement is True, and stops with an error if it is False
assert len(orders) == df['order_id'].nunique()
assert np.isclose(orders['revenue'].sum(), df['revenue'].sum())
assert np.isclose(orders['profit'].sum(), df['item_profit'].sum())

shipped_orders = orders[orders['is_shipped']]
print(f"Orders: {len(orders):,} | revenue ${orders['revenue'].sum():,.0f} | profit ${orders['profit'].sum():,.0f}")
print(f"Shipped orders: {len(shipped_orders):,} | late rate {shipped_orders['is_late'].mean():.1%}")
print(f"Fraud orders: {orders['is_fraud'].sum():,}")

Orders: 65,752 | revenue $33,054,402 | profit $3,966,903
Shipped orders: 62,897 | late rate 57.3%
Fraud orders: 1,488


In [28]:
orders['n_items'].value_counts().sort_index()

n_items
1    19850
2    11447
3    11398
4    11704
5    11353
Name: count, dtype: int64

## 3.14 Save outputs

In [29]:
df.to_csv('../data/clean/order_items_clean.csv', index=False)
orders.to_csv('../data/clean/orders_clean.csv', index=False)
pd.DataFrame(log).to_csv('../reports/data_quality_log.csv', index=False)
print('Saved', len(df), 'items,', len(orders), 'orders and', len(log), 'log entries')

Saved 180519 items, 65752 orders and 28 log entries


## Phase 3 summary
- 180,519 order items → 65,752 orders. 18 columns dropped with reasons; 35 kept and renamed; 10 derived columns added.
- Revenue = Order Item Total (after discount). Profit = Order Profit Per Order (item-level despite its name).
- Cancelled and suspected-fraud orders never shipped, so they are excluded from delivery analysis.
- Totals reconcile between the item and order tables and match the Phase 2 checks.
- Signs of simulated data are logged as limitations.
- Outputs: data/clean/order_items_clean.csv, data/clean/orders_clean.csv, reports/data_quality_log.md